enderio-base/src/main/java/com/enderio/base/common/capability/darksteel/DarkSteelUpgradeable.java

```diff
@@ -0,0 +1,173 @@
+import com.enderio.base.common.item.darksteel.upgrades.DarkSteelUpgradeRegistry;
+import com.enderio.base.common.item.darksteel.upgrades.EmpoweredUpgrade;
+import net.minecraft.nbt.CompoundTag;
+import net.minecraft.nbt.Tag;
+import net.minecraft.resources.ResourceLocation;
+import net.minecraft.world.item.ItemStack;
+
+import java.util.*;
+
+@SuppressWarnings("unused")
+public class DarkSteelUpgradeable implements IDarkSteelUpgradable {
+
+    //----------------------- Utils
+
+    public static ItemStack addUpgrade(ItemStack is, IDarkSteelUpgrade upgrade) {
+        is.getCapability(EIOCapabilities.DARK_STEEL_UPGRADABLE).ifPresent(upgradable -> upgradable.addUpgrade(upgrade));
+        return is;
+    }
+
+    public static Collection<IDarkSteelUpgrade> getUpgrades(ItemStack is) {
+        return is.getCapability(EIOCapabilities.DARK_STEEL_UPGRADABLE).map(IDarkSteelUpgradable::getUpgrades).orElse(Collections.emptyList());
+    }
+
+    public static boolean hasUpgrade(ItemStack is, String name) {
+        return is.getCapability(EIOCapabilities.DARK_STEEL_UPGRADABLE).map(upgradable -> upgradable.hasUpgrade(name)).orElse(false);
+    }
+
+    public static <T extends IDarkSteelUpgrade> Optional<T> getUpgradeAs(ItemStack is, String upgrade, Class<T> as) {
+        Optional<IDarkSteelUpgradable> cap = is.getCapability(EIOCapabilities.DARK_STEEL_UPGRADABLE).resolve();
+        if(cap.isEmpty()) {
+            return Optional.empty();
+        }
+        return cap.get().getUpgradeAs(upgrade, as);
+    }
+
+    public static Collection<IDarkSteelUpgrade> getUpgradesThatCanBeAppliedAtTheMoment(ItemStack is) {
+        return is.getCapability(EIOCapabilities.DARK_STEEL_UPGRADABLE).map(IDarkSteelUpgradable::getUpgradesThatCanBeAppliedAtTheMoment).orElse(Collections.emptyList());
+    }
+
+    public static Collection<IDarkSteelUpgrade> getAllPossibleUpgrades(ItemStack is) {
+        return is.getCapability(EIOCapabilities.DARK_STEEL_UPGRADABLE).map(IDarkSteelUpgradable::getAllPossibleUpgrades).orElse(Collections.emptyList());
+    }
+
+    //----------------------- Class
+
+    private final Map<String, IDarkSteelUpgrade> upgrades = new HashMap<>();
+
+    /** The type of item that is upgradable, used to determine valid upgrades.*/
+    private ResourceLocation onItem;
+
+    public DarkSteelUpgradeable(){
+        this(EnderIO.loc("empty"));
+    }
+
+    public DarkSteelUpgradeable(ResourceLocation onItem) {
+        this.onItem = onItem;
+    }
+
+    @Override
+    public void addUpgrade(IDarkSteelUpgrade upgrade) {
+        removeUpgradeInSlot(upgrade.getSlot());
+        upgrades.put(upgrade.getSerializedName(), upgrade);
+    }
+
+    @Override
+    public void removeUpgrade(String name) {
+        upgrades.remove(name);
+    }
+
+    private void removeUpgradeInSlot(String slot) {
+        Optional<String> toRemove = Optional.empty();
+        for(var entry : upgrades.entrySet()) {
+            if(entry.getValue().getSlot().equals(slot)) {
```

Comment:
```suggestion
        for (var entry : upgrades.entrySet()) {
            if (entry.getValue().getSlot().equals(slot)) {
```
add spaced before for/if

Reply:
pls check that for the other files too, because there are multiple places where this happened